src/mlpack/methods/dbscan/dbscan_impl.hpp

```diff
@@ -154,18 +155,71 @@ void DBSCAN<RangeSearchType, PointSelectionPolicy>::PointwiseCluster(
   std::vector<std::vector<size_t>> neighbors;
   std::vector<std::vector<double>> distances;
 
+  // Note that the strategy here is somewhat different from the original DBSCAN
+  // paper.  The original DBSCAN paper grows each cluster individually to its
+  // fullest extent; here, we use a UnionFind structure to grow each point into
+  // a local cluster (if it has enough points), and we combine with other local
+  // clusters.  The end result is the same.
+  //
+  // Define points as being either core points, or non-core points.  Core points
+  // have more than `minPoints` neighbors.  Non-core points are included into
+  // the first core point cluster that encounters them; if they are not included
+  // by anything, they are labeled as noise.
+  //
+  // We maintain a list of non-core points so that we can handle that logic
+  // correctly.
+
+  std::vector<bool> visited(data.n_cols, false);
+  std::vector<bool> nonCorePoints(data.n_cols, false);
+
   for (size_t i = 0; i < data.n_cols; ++i)
   {
     if (i % 10000 == 0 && i > 0)
       Log::Info << "DBSCAN clustering on point " << i << "..." << std::endl;
 
+    // Get the next index.
+    const size_t index = pointSelector.Select(i, data);
+    visited[index] = true;
+
     // Do the range search for only this point.
-    rangeSearch.Search(data.col(i), Range(0.0, epsilon), neighbors,
+    rangeSearch.Search(data.col(index), Range(0.0, epsilon), neighbors,
         distances);
 
-    // Union to all neighbors.
-    for (size_t j = 0; j < neighbors[0].size(); ++j)
-      uf.Union(i, neighbors[0][j]);
+    // Union to all neighbors if the point is not noise.
+    //
+    // Note that the +1 is needed because our range search will return the point
+    // itself as within the range.
+    //
+    // If the point is noise, we leave its label as undefined (i.e. we do no
+    // unioning).
+    if (neighbors[0].size() >= minPoints + 1)
```

Comment:
I think +1 will not be there as choosen point will also lie inside the circle of eps acorss that point.
For the datset - 
	1	2	2	8	8	25
	2	2	3	7	8	80
and eps=3 and minpts=2.
This shows label -
        0      0      0      SIZE_MAX      SIZE_MAX      SIZE_MAX      
But actual label is -
        0      0      0      1     1     SIZE_MAX 

Example SRC - First example in https://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html

Reply:
Thanks, you are correct, and you saw exactly why I misunderstood.  Using the reasoning in your footnote, I removed the `+ 1` throughout in 8e1f30ba58.  Note that this means for the dual-tree search, we use `- 1` now, as dual-tree monochromatic range search will not return the point in its own range (similar to k-nearest-neighbor search).